File: deepgoal/core/types/hooks.py
from __future__ import annotations

from enum import Enum
from typing import Any, Literal, Protocol

from pydantic import BaseModel


class HookEvent(str, Enum):
    PRE_TOOL_USE = "PreToolUse"
    POST_TOOL_USE = "PostToolUse"
    POST_TOOL_USE_FAILURE = "PostToolUseFailure"
    USER_PROMPT_SUBMIT = "UserPromptSubmit"
    STOP = "Stop"
    SUBAGENT_STOP = "SubagentStop"
    SUBAGENT_START = "SubagentStart"
    NOTIFICATION = "Notification"
    PRE_COMPACT = "PreCompact"
    PERMISSION_REQUEST = "PermissionRequest"


class HookResult(BaseModel):
    continue_: bool = True
    decision: Literal["block"] | None = None
    reason: str | None = None
    additional_context: str | None = None


class HookHandler(Protocol):
    async def __call__(
        self, event: HookEvent, payload: dict[str, Any]
    ) -> HookResult: ...

File: deepgoal-ts/src/core/types/hooks.ts
import type { JsonObject } from './common.js';

export enum HookEvent {
  PreToolUse = 'PreToolUse',
  PostToolUse = 'PostToolUse',
  PostToolUseFailure = 'PostToolUseFailure',
  UserPromptSubmit = 'UserPromptSubmit',
  Stop = 'Stop',
  SubagentStop = 'SubagentStop',
  SubagentStart = 'SubagentStart',
  Notification = 'Notification',
  PreCompact = 'PreCompact',
  PermissionRequest = 'PermissionRequest',
}

export interface HookPayload {
  data: JsonObject;
}

export interface HookResult {
  continue: boolean;
  decision?: 'block' | undefined;
  reason?: string | undefined;
  additionalContext?: string | undefined;
}

export type HookHandler = (event: HookEvent, payload: HookPayload) => Promise<HookResult>;

export function createHookPayload(data: JsonObject = {}): HookPayload {
  return { data };
}

export function createHookResult(result: Partial<HookResult> = {}): HookResult {
  return {
    continue: result.continue ?? true,
    decision: result.decision,
    reason: result.reason,
    additionalContext: result.additionalContext,
  };
}
